src/ProcessManager.php

```diff
@@ -721,59 +745,54 @@ protected function bootstrapFailed($port)
      */
     protected function checkChangedFiles($restartSlaves = true)
     {
-        if ($this->inReload) {
+        if ($this->inChangesDetectionCycle) {
             return false;
         }
 
-        clearstatcache();
-
-        $reload = false;
-        $filePath = '';
         $start = microtime(true);
+        $hasChanged = false;
+
+        $this->inChangesDetectionCycle = true;
+
+        clearstatcache();
 
-        foreach ($this->filesToTrack as $idx => $filePath) {
+        foreach ($this->filesLastMTime as $filePath => $knownMTime) {
             if (!file_exists($filePath)) {
                 continue;
             }
 
-            $currentFileMTime = filemtime($filePath);
-
-            if (isset($this->filesLastMTime[$filePath])) {
-                if ($this->filesLastMTime[$filePath] !== $currentFileMTime) {
-                    $this->filesLastMTime[$filePath] = $currentFileMTime;
-
-                    $md5 = md5_file($filePath);
-                    if (!isset($this->filesLastMd5[$filePath]) || $md5 !== $this->filesLastMd5[$filePath]) {
-                        $this->filesLastMd5[$filePath] = $md5;
-                        $reload = true;
-
-                        //since chances are high that this file will change again we
-                        //move this file to the beginning of the array, so next check is way faster.
-                        unset($this->filesToTrack[$idx]);
-                        array_unshift($this->filesToTrack, $filePath);
-                        break;
-                    }
-                }
-            } else {
-                $this->filesLastMTime[$filePath] = $currentFileMTime;
+            if ($knownMTime !== filemtime($filePath) && $this->filesLastMd5[$filePath] !== md5_file($filePath, true)) {
+                $this->output->writeln(
+                    sprintf("<info>[%s] File %s has changed.</info>", date('d/M/Y:H:i:s O'), $filePath)
+                );
+                $hasChanged = true;
+                break;
             }
         }
 
-        if ($reload && $restartSlaves) {
+        if ($hasChanged) {
             $this->output->writeln(
                 sprintf(
-                    "<info>[%s] File changed %s (detection %.3f, %d). Reloading workers.</info>",
+                    "<info>[%s] At least one of %u known files was changed.</info>",
```

Comment:
We should add the count of files that were changed and if only one, we should print the path of that one file as before. Makes things easier for development.

Reply:
> We should add the count of files that were changed

Now we're breaking the changes detection loop when the first changed file was found (I guess for avoiding excess disk I/O operations). Should we sacrifice it to add changes count to the log?

Reply:
@marcj, what do you think about it?

Reply:
:bell: ding!

Reply:
> We should add the count of files that were changed and if only one, we should print the path of that one file as before. Makes things easier for development.

@marcj we didn't do this before. Instead we're just restarting on the first change observed which is also why the changed file it put on top of the change list. Imho this should stay as it was- no tracking on number of changed files.

Reply:
ok for now, we should however increase developer experience in the future and print what file has been changed 👍